File: src/lib/api/plans.ts
import type { Plan, CreatePlanParams, UpdatePlanParams } from "@/lib/types";
import { invoke, withTauriError } from "./utils";

export async function getPlan(id: string): Promise<Plan> {
  return withTauriError("获取 Plan", async () => {
    return invoke<Plan>("get_plan", { id });
  });
}

export async function getPlans(): Promise<Plan[]> {
  return withTauriError("获取 Plan 列表", async () => {
    return invoke<Plan[]>("get_plans");
  });
}

export async function createPlan(data: CreatePlanParams): Promise<Plan> {
  return withTauriError("创建 Plan", async () => {
    return invoke<Plan>("create_plan", {
      title: data.title,
      description: data.description || null,
      startDate: data.start_date || null,
      endDate: data.end_date || null,
    });
  });
}

export async function updatePlan(
  id: string,
  data: UpdatePlanParams,
): Promise<Plan> {
  return withTauriError("更新 Plan", async () => {
    return invoke<Plan>("update_plan", {
      id,
      title: data.title,
      description: data.description,
      startDate: data.start_date,
      endDate: data.end_date,
      status: data.status,
    });
  });
}

export async function deletePlan(id: string): Promise<void> {
  return withTauriError("删除 Plan", async () => {
    return invoke<void>("delete_plan", { id });
  });
}

File: src-tauri/src/commands/plans.rs
// Plan CRUD commands

use crate::commands::repositories::PlanRepository;
use crate::log_command;
use crate::models::Plan;
use crate::AppState;

#[tauri::command]
pub fn get_plan(state: tauri::State<AppState>, id: String) -> Result<Plan, String> {
    log_command!("get_plan", {
        let conn = state.db.lock().map_err(|e| e.to_string())?;
        PlanRepository::get_by_id(&conn, &id)
    })
}

#[tauri::command]
pub fn get_plans(state: tauri::State<AppState>) -> Result<Vec<Plan>, String> {
    log_command!("get_plans", {
        let conn = state.db.lock().map_err(|e| e.to_string())?;
        PlanRepository::get_all(&conn)
    })
}

#[tauri::command]
pub fn create_plan(
    state: tauri::State<AppState>,
    title: String,
    description: Option<String>,
    start_date: Option<String>,
    end_date: Option<String>,
) -> Result<Plan, String> {
    log_command!("create_plan", {
        let conn = state.db.lock().map_err(|e| e.to_string())?;
        let id = uuid::Uuid::new_v4().to_string();
        PlanRepository::create(
            &conn,
            &id,
            &title,
            description.as_deref(),
            start_date.as_deref(),
            end_date.as_deref(),
        )
    })
}

#[tauri::command]
pub fn update_plan(
    state: tauri::State<AppState>,
    id: String,
    title: Option<String>,
    description: Option<String>,
    start_date: Option<String>,
    end_date: Option<String>,
    status: Option<String>,
) -> Result<Plan, String> {
    log_command!("update_plan", {
        let conn = state.db.lock().map_err(|e| e.to_string())?;
        PlanRepository::update(
            &conn,
            &id,
            title.as_deref(),
            description.as_deref(),
            start_date.as_deref(),
            end_date.as_deref(),
            status.as_deref(),
        )
    })
}

#[tauri::command]
pub fn delete_plan(state: tauri::State<AppState>, id: String) -> Result<(), String> {
    log_command!("delete_plan", {
        let conn = state.db.lock().map_err(|e| e.to_string())?;
        conn.execute("DELETE FROM tasks WHERE plan_id = ?", [&id])
            .map_err(|e| e.to_string())?;
        PlanRepository::delete(&conn, &id)
    })
}

#[tauri::command]
pub fn update_plan_sort_order(
    state: tauri::State<AppState>,
    id: String,
    sort_order: i32,
) -> Result<(), String> {
    log_command!("update_plan_sort_order", {
        let conn = state.db.lock().map_err(|e| e.to_string())?;
        PlanRepository::update_sort_order(&conn, &id, sort_order)
    })
}

#[tauri::command]
pub fn reorder_plans(
    state: tauri::State<AppState>,
    orders: Vec<(String, i32)>,
) -> Result<usize, String> {
    log_command!("reorder_plans", {
        let conn = state.db.lock().map_err(|e| e.to_string())?;
        PlanRepository::reorder(&*conn, &orders)
    })
}
